### ISSUE TITLE
can not boot after switching to linux-lts, linux-lts-headers and archzfs-linux-lts

### ISSUE THREAD

**Issue by stevleibelt** [archzfs/archzfs#77]
Hey,

After waiting for zfsonlinux and releaseing the next package that is supporting the linux 4.7.x, I thought it is a good idea to switch to the linux-lts kernel.
What I did was

```sh
pacman -Ss
pacman -S linux-lts linux-lts-headers
pacman -R archzfs-linux linux linux-headers
pacman -S archzfs-linux-lts
#have to do this to prevent error "failed to get canonical path of"
ln -s /dev/mapper/crypt-br01 /dev/crypt-br01
grub-mkconfig -o /boot/grub/grub.cfg
```
I have some kind of special system setting which I will explain in the following line. I am simple following my own [notes](https://github.com/stevleibelt/General_Howtos/blob/master/operation_system/linux/arch/installOnZfsWithLuks.md).
My primary disk is splitted into three partitions.
sda1 is a bios partition
sda2 is the boot partition
sda3 is the zfs root secured with luks

The mkinitcpio was running without any issue I could spot.
What happens now/after a reboot is the following console output.

```sh
starting version 231
ash: 1: unknown operand
ERROR: Unable to find root devide `ZFS=br01/`
You are dropped to a recovery shell
Type `exit` to try and continue booting
sh: can`t access tty: job control turned off
[rootfs ]#
```

When I try to make a "zpool status" I get the message "The ZFS modules are not loaded. Try running `/sbin/modprobe zfs`.

I tried to follow [this howto](https://github.com/archzfs/archzfs/issues/76#issuecomment-246949698).

```sh
/sbin/modprobe zfs
cryptsetup luksOpen /dev/sda3 crypt-br01
zpool import -R /new_root br01
exec /sbin/switch_root /new_root /sbin/init
```

Which results in the following output.

```sh
Trying to run as user instance, but the system has not been bootet with systemd.
/init: line 48: mount: not found
You are now being dropped into an emergency shell:
/init: line 48: can`t open /dev/console: no such file
/init: line 68: stat: not found
ERROR: Failed to mount the real root device.
Bailing out, you are on your own. Good luck

...

Kernel panic ...
```

I must admint, my force is not strong enough or have reached its limit. I have archiso usb stick with me (with installed archzfs), so I can do some stuff. My problem is so far, I have no idea what to do.
Switching back to the latest linux and archzfs-linux is currently not possible since there is already a new kernel around which is not supported by archzfs-linux right now.

Could you give me any hint how I can fix this issue? Why is the modprobe for zfs not working? There is an entry in `/etc/modules-load.d`. Is switching to the lts kernel from the regular kernel such a bad idea?
Thanks in advance,
stev

[edit: 2016-09-16: 23:09, Europe/Berlin Time]

Also doing the follwing steps after the initial boot does not work.

```sh
/sbin/modprobe zfs
cryptsetup luksOpen /dev/sda3 crypt-br01
exit
```

Results in the output.

```sh
filesystem `ZFS=br01/`cannot be mounted, unable to open the dataset
...
```
Since there is nothing I can break so far, I did `zpool import br01` which results in the following output.

```sh
cannot mount `/`: directory is not empty
exit
no pools availabe to import
ERROR: Failed to mount the real root device.
Bailing out, you are on your own. Good luck.
...
```

I think this is it for today. I am off to go to bed. Again, thanks for any kind of response.

**Comment by demizer** [comment]
Thank you for the detailed issue report. I don't yet have any tests that use grub or encryption and will have to add regression tests for this, but I am currently short on spare time.

```
ash: 1: unknown operand
```
Is cause bit the ash interpreter parsing the initcpio hook I believe. I will see if there is a better way to debug this.

**Comment by demizer** [comment]
Well crap. Stupid me, ash is used in busy box, not bash. I've been writing the initscripts like bash scripts. I will have to lint the initscripts to make sure ash can parse them. The manual for ash is not very good ... http://linux.die.net/man/1/ash

**Comment by stevleibelt** [comment]
Hi @demizer 

thanks for your time. Don't feel stressed and do your stuff first. I am not mad or anything else, especially since I know what kind of favour you do for us archzfs users.
If I am mad, than only against myself since I my kungfu is not strong enough to solve this issue on my own :-D (infect, I was able to boot my system once once yesterday evening but I have no idea how I did this - I could not reproduce the booting :-().

A information I missed to add is the fact that I did the downgrade from 4.7.2-1-ARCH (output of uname -r).
I've also tried to do adding a softlink of the zfs.target on my own when working with my created [archzfs iso](https://github.com/stevleibelt/arch-linux-live-cd-iso-with-zfs/). Within this iso, I only have the latest available and supported kernel (which is 4.7.2-1-ARCH also).
What I did when adding a softlink was:

```sh
ln -s /usr/lib/systemd/system/zfs.target /etc/systemd/system/multi-user.target.wants/zfs.target
```

There is also a zfs.conf inside /etc/modules-load.d/.

For the ash error message, I have this for ... lets call it ages, and on multiple machines (I am running 4 with the disk layout), so this isn't an urgent issue for me.

**Comment by stevleibelt** [comment]
@demizer, I have fixed it and I guess this is nothing you could fix from your side :-). Thanks for your quick response :+1: 

I had an older version of my grub.cfg backup-ed $somewhere so I could do a vimdiff :-).

Most important line change is.

```sh
 #generated entry by grub-mkconfig
linux /vmlinuz-linux-lts root=ZFS=br01/ rw quiet
#to this line
linux /vmlinuz-linux-lts zfs=br01 rw cryptodevice=/dev/disk/by-uuid/<uuid>:crypto-br01 quiet
```

System is up and running again, repeatable :-D.

Enjoy the day.
